File: golang/multithread_sort/merge_sort.go
package multithread_sort

import (
	"algoritmos-tcc-golang/constants"
	"sync"
)

type mergeSort struct{}

// Array len % 4 == 0
func NewMultithreadMergeSort() *mergeSort {
	return &mergeSort{}
}

func (ss *mergeSort) Sort(array []int) {
	arrayLen := len(array)
	var wg sync.WaitGroup

	for i := 0; i < constants.THREADS_NUMBER; i++ {
		low := i * (arrayLen / 4)
		high := (i+1)*(arrayLen/4) - 1
		mid := low + (high-low)/2

		wg.Add(1)
		go func() {
			defer wg.Done()

			ss.mergeSort(array, low, mid)
			ss.mergeSort(array, mid+1, high)

			ss.merge(array, low, mid, high)
		}()
	}

	wg.Wait()

	ss.merge(array, 0, (arrayLen/2-1)/2, arrayLen/2-1)
	ss.merge(array, arrayLen/2, arrayLen/2+(arrayLen-1-arrayLen/2)/2, arrayLen-1)
	ss.merge(array, 0, (arrayLen-1)/2, arrayLen-1)
}

func (ss *mergeSort) mergeSort(arr []int, low int, high int) {
	if low < high {
		mid := low + (high-low)/2

		ss.mergeSort(arr, low, mid)
		ss.mergeSort(arr, mid+1, high)

		ss.merge(arr, low, mid, high)
	}
}

func (ss *mergeSort) merge(arr []int, low int, mid int, high int) {
	var i, j, k int
	n1 := mid - low + 1
	n2 := high - mid

	left := make([]int, n1)
	right := make([]int, n2)

	for i = 0; i < n1; i++ {
		left[i] = arr[low+i]
	}

	for j = 0; j < n2; j++ {
		right[j] = arr[mid+1+j]
	}

	i = 0
	j = 0
	k = low
	for i < n1 && j < n2 {
		if left[i] <= right[j] {
			arr[k] = left[i]
			i++
		} else {
			arr[k] = right[j]
			j++
		}
		k++
	}

	for i < n1 {
		arr[k] = left[i]
		k++
		i++
	}

	for j < n2 {
		arr[k] = right[j]
		k++
		j++
	}
}

File: cpp/multithread_sort/merge_sort.cpp
namespace multithread_merge_sort
{
  using namespace std;
  using namespace constants;

  void merge(vector<int> *array, int low, int mid, int high)
  {
    int i, j, k;
    int n1 = mid - low + 1;
    int n2 = high - mid;

    vector<int> *left = new vector<int>;
    vector<int> *right = new vector<int>;

    for (i = 0; i < n1; i++)
    {
      left->push_back(array->operator[](low + i));
    }

    for (i = 0; i < n2; i++)
    {
      right->push_back(array->operator[](mid + 1 + i));
    }

    i = 0;
    j = 0;
    k = low;
    while (i < n1 && j < n2)
    {
      if (left->operator[](i) <= right->operator[](j))
      {
        array->operator[](k) = left->operator[](i);
        i++;
      }
      else
      {
        array->operator[](k) = right->operator[](j);
        j++;
      }
      k++;
    }

    while (i < n1)
    {
      array->operator[](k) = left->operator[](i);
      k++;
      i++;
    }

    while (j < n2)
    {
      array->operator[](k) = right->operator[](j);
      k++;
      j++;
    }
  }

  void mergeSort(vector<int> *array, int low, int high)
  {
    if (low < high)
    {
      int mid = low + (high - low) / 2;

      mergeSort(array, low, mid);
      mergeSort(array, mid + 1, high);

      merge(array, low, mid, high);
    }
  }

  void thread_sort(vector<int> *array, int low, int mid, int high)
  {
    mergeSort(array, low, mid);
    mergeSort(array, mid + 1, high);

    merge(array, low, mid, high);
  }

  void sort(vector<int> *array)
  {
    int arrayLen = array->size();
    vector<thread> threads;

    for (int i = 0; i < THREADS_NUMBER; i++)
    {
      int low = i * (arrayLen / 4);
      int high = (i + 1) * (arrayLen / 4) - 1;
      int mid = low + (high - low) / 2;

      thread t(thread_sort, array, low, mid, high);

      threads.push_back(move(t));
    }

    for (thread &t : threads)
    {
      t.join();
    }

    merge(array, 0, (arrayLen / 2 - 1) / 2, arrayLen / 2 - 1);
    merge(array, arrayLen / 2, arrayLen / 2 + (arrayLen - 1 - arrayLen / 2) / 2, arrayLen - 1);
    merge(array, 0, (arrayLen - 1) / 2, arrayLen - 1);
  }
}
